[ci2] add heal logic

cseed opened the pull request:
stacked on 3 other ci2 PRs.

 - add heal logic
 - add job log endpoint
 - some simplifications: PRs have a watched branch target (which has a sha) and a source sha, but don't need the source branch.  Got rid of FQSHA.
 - Ran by hand, seems to work.  Will add tests back once build is actually doing something non-trivial.

danking commented on ci2/Makefile:
```diff
@@ -30,7 +30,7 @@ build:
 
 run-local:
-	python3 ci/ci.py
+	PYTHONPATH=../batch python3 ci/ci.py
```
this should probably be `PYTHONPATH=../batch:$PYTHONPATH`, no? How does python find the other installed libraries?

danking commented on ci2/ci/ci.py:
```diff
@@ -1,16 +1,107 @@
 routes = web.RouteTableDef()
 
+@routes.get('/')
+@aiohttp_jinja2.template('index.html')
```
I don't understand why GitHub is showing this if it's already merged?

cseed replied:
It needs to be rebased.

cseed commented:
OK, I cleaned this up a bit.

Now stacked on: https://github.com/hail-is/hail/pull/5826

Summary of changes:
 - added heal
 - pipeline is now batch rather than job centeric
 - batch logs page shows logs for all batch jobs
 - GET /batches/{id} endpoint now returns entire array of jobs, instead of the state counter and exit_codes

cseed commented:
@danking rebased, no longer stacked, this should be ready for a look.  Particularly interested in your take on the heal functions after our discussions.

danking commented:
@cseed unrelated but I think you've configured your local git email and name to be hail-ci-leader

danking commented on ci2/ci/github.py:
```diff
@@ -178,6 +139,45 @@ def from_gh_json(gh_json):
+                    })
+
+                # FIXME build
+                await self.batch.create_job('alpine', ['echo', 'foo'])
```
we should be setting the attributes here yeah?

cseed replied:
I don't need them.  I only look up the batch.  I don't think the CI will ever be concerned with individual jobs (except for logging, etc.)

danking commented:
I think you'll need the source branch eventually so that you can fetch and merge it into master

cseed commented on ci2/ci/github.py:
```diff
@@ -190,13 +190,45 @@ def __init__(self, branch):
             number = gh_json_pr['number']
-            pr = PR.from_gh_json(gh_json_pr)
-            await pr.refresh(gh)
+            if self.prs is not None and number in self.prs:
```
I tried to be careful about using None to represent unknown.  In particular, I didn't want to start with {} and then have ci blow away all of the running jobs because it doesn't think there are any PRs.

cseed commented on ci2/ci/github.py:
```diff
@@ -178,6 +139,45 @@ def from_gh_json(gh_json):
 
         self.state = total_state
```
Yep, totally agree.  Also, the loop computing latest_state_by_login wasn't setting 'DISMISSED'.  I fixed that, too.

cseed commented on ci2/ci/github.py:
```diff
@@ -90,75 +90,36 @@ def to_dict(self):
     async def refresh(self, gh):
         latest_state_by_login = {}
-        async for review in gh.getiter(f'/repos/{self.target.branch.repo.short_str()}/pulls/{self.number}/reviews'):
+        async for review in gh.getiter(f'/repos/{self.target_branch.branch.repo.short_str()}/pulls/{self.number}/reviews'):
```
Yeah, it is only called from a watched branch, which just did a get pulls and called update_from_gh_json, so this function doesn't need to update the pull itself.  I renamed it to `refresh_review_state` as suggested.

cseed commented:
Yeah, in the next PR I add PR.source_branch, but I didn't find a need for FQSHA.

danking commented on ci2/ci/ci.py:
```diff
@@ -40,40 +48,61 @@
 
 async def on_startup(app):
-    asyncio.ensure_future(refresh_loop())
+    asyncio.ensure_future(refresh_loop(app))
```
ah I didn't realize you could do this. I fixed up batch to async everything except the k8s watch, but I incorporated other coroutines like this:
```python
    try:
        app.add_routes(routes)
        loop.run_until_complete(asyncio.gather(
            _run_app(app, host='0.0.0.0', port=port),
            polling_event_loop(loop)))
    finally:
        _cancel_all_tasks(loop)
        loop.run_until_complete(loop.shutdown_asyncgens())
```

I think your approach here is probably better.

cseed replied:
One difference is you try to clean up here (cancel tasks, shutdown_asyncgens), and I don't.  I don't know if that's important?  Maybe I should record the refresh loop and try to cancel it on shutdown?

danking replied:
`app.run_app(...)` includes clean up logic for you

danking commented on ci2/ci/github.py:
```diff
@@ -190,13 +189,45 @@ def __init__(self, branch):
         self.prs = new_prs
+
+        for pr in self.prs.values():
+            await pr.refresh_review_state(gh)
```
ok, I think there's still one missing piece of logic: where do we remove a pr that has been closed? It seems like we should track the numbers we saw in the `async for` loop.  After the loop, we should:
```
for k in self.prs.keys() - seen_pr_numbers:
    del self.prs[k]
```

cseed replied:
I reassign prs just before this:

 >         self.prs = new_prs

danking replied:
👍
